drivers/input/input_stmpe811.c

```diff
@@ -0,0 +1,481 @@
+	if (err < 0) {
+		return err;
+	}
+
+	/* Touch Screen Control config */
+	err += i2c_reg_write_byte_dt(&config->bus, STMPE811_TSC_CFG_REG,
+				     stmpe811_tsc_config_bits(dev));
+	if (err < 0) {
+		return err;
+	}
+
+	/* Configure the touch FIFO threshold */
+	err += i2c_reg_write_byte_dt(&config->bus, STMPE811_FIFO_TH_REG, FIFO_TH_SINGLE_POINT);
+	if (err < 0) {
+		return err;
+	}
+
+	/* Clear the FIFO memory content */
+	err += i2c_reg_write_byte_dt(&config->bus, STMPE811_FIFO_STA_REG, FIFO_STA_CLEAR);
+	if (err < 0) {
+		return err;
+	}
+
+	/* Put the FIFO back into operation mode */
+	err += i2c_reg_write_byte_dt(&config->bus, STMPE811_FIFO_STA_REG, FIFO_STA_OPERATIONAL);
+	if (err < 0) {
+		return err;
+	}
+
+	/* Set the range and accuracy of the pressure measurement (Z) */
+	err += i2c_reg_write_byte_dt(&config->bus, STMPE811_TSC_FRACT_XYZ_REG,
+				     PRESSURE_MEASUREMENT_CONF);
+	if (err < 0) {
+		return err;
+	}
+
+	/* Set the driving capability (limit) of the device for TSC pins */
+	err += i2c_reg_write_byte_dt(&config->bus, STMPE811_TSC_I_DRIVE_REG, TSC_I_DRIVE_LIMIT);
+	if (err < 0) {
+		return err;
+	}
+
+	/* Touch screen control configuration */
+	err += i2c_reg_write_byte_dt(&config->bus, STMPE811_TSC_CTRL_REG, TSC_CONF);
+	if (err < 0) {
+		return err;
+	}
+
+	/**
+	 * Clear all the status pending bits if any.
+	 * Writing '1' to this register clears the corresponding bits.
+	 * This is an 8-bit register, so writing 0xFF clears all.
+	 */
+	err += i2c_reg_write_byte_dt(&config->bus, STMPE811_INT_STA_REG, BIT_MASK(8));
+	k_msleep(STMPE811_WAIT_DELAY_MS);
+
+	return err;
+}
+
+static bool stmpe811_ts_detect_touch(const struct device *dev)
+{
+	const struct stmpe811_config *config = dev->config;
+	uint8_t fifo_level;
+
+	/* Read FIFO level */
+	i2c_reg_read_byte_dt(&config->bus, STMPE811_FIFO_SIZE_REG, &fifo_level);
+
+	return fifo_level;
+}
+
+static int stmpe811_ts_get_state(const struct device *dev)
+{
+	const struct stmpe811_config *config = dev->config;
+	struct stmpe811_data *data = dev->data;
+
+	bool touch_detected = stmpe811_ts_detect_touch(dev);
+	uint8_t data_xy[3];
+	uint32_t uldata_xy;
+
+	if (touch_detected > 0) {
+		int ret = i2c_burst_read_dt(&config->bus, STMPE811_TSC_DATA_NON_INC_REG, data_xy,
+					    sizeof(data_xy));
+		if (ret < 0) {
+			return ret;
+		}
+
+		data->touch_detected = 1;
+		/* Calculate positions values */
+		uldata_xy = (data_xy[0] << 16) | (data_xy[1] << 8) | data_xy[2];
+		data->touch_x = (uldata_xy >> 12U) & BIT_MASK(12);
+		data->touch_y = uldata_xy & BIT_MASK(12);
+
+		/* Reset FIFO */
+		ret = i2c_reg_write_byte_dt(&config->bus, STMPE811_FIFO_STA_REG, FIFO_STA_CLEAR);
+		if (ret < 0) {
+			return ret;
+		}
+
+		ret = i2c_reg_write_byte_dt(&config->bus, STMPE811_FIFO_STA_REG,
+					    FIFO_STA_OPERATIONAL);
+		if (ret < 0) {
+			return ret;
+		}
+	} else if (touch_detected == 0) {
+		data->touch_detected = 0;
+	} else {
+		return -EINVAL;
+	}
```

Comment:
it's usually a good idea to keep the normal execution code inline and the validation first, how about:

```c
if (touch_detected <0) {
        return -EINVAL;
}

if (touch_detected == 0) {
        data->touch_detected = 0;
        return 0;
}

...rest of the code...
```